5주차 - 고양이 장난감가게 만들기(심화)

eastshine-high opened the pull request:
안녕하세요! 조금 헤매어서 늦게 커밋했네요.
잘 부탁드립니다.

1차 커밋 내용.
구현
DozzerMapper를 이용한 사용자 생성(create)

미구현
create를 제외한 기능들(update, delete)

eastshine-high committed:
고양이 장난감가게 만들기(심화) 1차 커밋

구현 내용
DozzerMapper를 이용한 사용자 생성(Create)

미구현 내용
create를 제외한 기능들(update, delete)

johngrib commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,15 +1,21 @@
+
+    // 스프링이 관리하는 객체를 만들 수 있다.
+    // 매번 만드는 것이 아니라 싱글턴으로 만들어 놓고 의존성 주입을 하여 사용.
+    @Bean
```
`@Bean`의 javadoc을 읽고 내용을 요약해 설명해 보세요. 그동안 javadoc 작성 연습을 한 것은 javadoc을 읽기 위한 훈련이기도 했어요.

johngrib commented on app/build.gradle:
```diff
@@ -51,6 +51,10 @@ dependencies {
     annotationProcessor 'org.projectlombok:lombok:1.18.16'
 
+    //DozerMapper
+    implementation('com.github.dozermapper:dozer-core:6.4.0')
```
dozer mapper의 기능과 목적에 대해 소개해 보세요. 공식 문서 링크도 같이 있다면 좋겠죠.

eastshine-high replied:
블로그에 정리한 내용을 링크합니다.
https://eastshine.notion.site/Dozer-mapper-de1b6a6e0c704755a78934a67003da6b

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/User.java:
```diff
@@ -0,0 +1,31 @@
+
+@Getter
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
```
이 두 애노테이션은 위험한 애노테이션으로 유명합니다. `@AllArgsConstructor`의 경우 사용을 금지하는 회사들도 많은 편이에요. 그 이유를 조사해서 설명해 보세요.

eastshine-high replied:
제거하였습니다. 이유를 조사하는대로 다시 커멘트 작성하겠습니다.

eastshine-high replied:
슬랙에 올려주신 내용을 확인하고 UserData.java에서는 AllArgsConstructor를 제거해 주었었는데 User.java부분을 남겨 두었었네요ㅜㅜ
이유는 생성자 argument의 순서가 바뀔 경우 치명적인 오류를 발생시킬 가능성을 내포하고 있어서 그렇습니다.
https://kwonnam.pe.kr/wiki/java/lombok/pitfall
위의 내용을 참고하였습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/dto/UserData.java:
```diff
@@ -0,0 +1,38 @@
+
+import javax.validation.constraints.NotBlank;
+
+@Setter
```
`@Setter`가 없어도 괜찮지 않을까요? 실제 업무에서는 `setter`를 거의 사용하지 않습니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -0,0 +1,59 @@
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+
+class UserServiceTest {
```
테스트를 계층 구조화해서 체계있게 설명을 작성해 보세요.

eastshine-high committed:
2차 커밋

구현 내용
- 계층 테스트를 이용한 UserService 기능 구현

- 1차 커밋 피드백 반영
@Setter ,@AllArgsConstructor, @NoArgsConstructor 제거

미구현 내용
UserController

eastshine-high commented:
2차 커밋
구현 내용
- 계층 테스트를 이용한 UserService 기능 구현

- 1차 커밋 피드백 반영
@Setter ,@AllArgsConstructor, @NoArgsConstructor 제거

미구현 내용
UserController

johngrib commented on app/src/main/java/com/codesoom/assignment/dto/UserData.java:
```diff
@@ -4,16 +4,14 @@
 
-@Setter
 @Getter
 @Builder
```
`@Builder`를 클래스 시그니처 위에 선언하는 것은 권장되지 않습니다. 그 이유에 대해 조사해 보세요.

eastshine-high replied:
Builder를 직접 작성한 생성자에 선언하였습니다.

Builder를 클래스 시그니처 위에 선언하는 것은 AllArgsConstructor를 사용하는 것을 내포합니다.
AllArgsConstructor의 위험성은 위의 커멘트에 설명하였습니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -33,27 +54,158 @@ void setUpAll() {
         });
     }
 
+    @Nested
+    class Describe_createUser {
+
+        @Test
+        void it_returns_a_cretaed_user() {
```
`@DisplayName`으로 테스트 설명문을 작성해 보세요. Java에서는 클래스 이름과 메소드 이름만으로는 충분히 설명하기 어렵습니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -33,27 +54,158 @@ void setUpAll() {
+                id = INVALID_ID;
+            }
 
+            @Test
+            void it_throws_UserNotFoundException() {
+                assertThatThrownBy(() -> assertThat(userService.getUser(id)))
+                        .isInstanceOf(UserNotFoundException.class);
+            }
```
이 테스트는 좋지 않습니다. 리포지토리 구현이 바뀌어서 `UsernotFoundException`이 아닌 다른 예외를 던지게 되어도 이 테스트는 성공할 것입니다.

eastshine-high replied:
repositoryd에서 던지지 않는 예외를 던지는 given문을 사용했네요.
위의 테스트 setUpAll에서 
given(userRepository.findById(INVALID_ID)).willThrow(UserNotFoundException.class);
다음의 문장을 삭제해 주었습니다.
이를 통해 잘못된 테스트의 위험성을 알게되었습니다.

eastshine-high committed:
3차 커밋

구현 내용
UserController 및 2차 커밋에 대한 피드백 반영

미구현 내용
UserServiceTest 일부 DisplayName 설명

johngrib reviewed (commented):
한 주간 고생 많으셨습니다. 혹시 과제 진행이 순탄하지 않으셨나요? 또는 다른 어려움이 있지는 않으셨나요?

eastshine-high commented:
> 한 주간 고생 많으셨습니다. 혹시 과제 진행이 순탄하지 않으셨나요? 또는 다른 어려움이 있지는 않으셨나요?

eastshine-high commented:
벌초를 다녀오느라 커밋이 조금 늦었네요ㅜㅜ

3차 커밋 내용입니다.

구현 내용
UserController 및 2차 커밋에 대한 피드백 반영

미구현 내용
UserServiceTest 일부 DisplayName 설명
